feat: Migrate to JUnit 5 and add parallel test execution

AbgarSim opened the pull request:
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/java-bigquery/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Fixes googleapis/google-cloud-java#12142  ☕️


Before Optimisitation
<img width="653" height="182" alt="Screenshot 2025-12-19 at 12 03 10" src="https://github.com/user-attachments/assets/e06b6ee8-0ef1-44b2-acd3-7c1fbdd41124" />

After Optimisation
<img width="671" height="184" alt="Screenshot 2025-12-19 at 11 57 23" src="https://github.com/user-attachments/assets/a39ab436-e3bf-409e-bf66-4f3d26083ddd" />


If you write sample code, please follow the [samples format](
https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/SAMPLE_FORMAT.md).

AbgarSim commented on google-cloud-bigquery/src/test/java/com/google/cloud/bigquery/JobTest.java:
```diff
@@ -18,14 +18,9 @@
-import static org.junit.Assert.assertSame;
-import static org.junit.Assert.assertTrue;
-import static org.junit.Assert.fail;
+import static org.junit.Assert.*;
```
Done

AbgarSim commented on google-cloud-bigquery/src/test/java/com/google/cloud/bigquery/InsertAllRequestTest.java:
```diff
@@ -17,14 +17,15 @@
 package com.google.cloud.bigquery;
 
 import static com.google.common.truth.Truth.assertThat;
+import static org.junit.Assert.assertThrows;
```
Done

lqiu96 commented on pom.xml:
```diff
@@ -134,6 +134,12 @@
+      <dependency>
+        <groupId>org.mockito</groupId>
+        <artifactId>mockito-junit-jupiter</artifactId>
+        <version>2.17.0</version>
```
IIUC, this is part of the migration to JUnit 5 where the `@ExtendWith(MockitoExtension.class)` annotations are split out to an extension package.

I think this is fine as I also noticed that we don't manage mockito v5 upstream so I think it should be fine to override it here for now. qq, I don't think this is the latest version: https://central.sonatype.com/artifact/org.mockito/mockito-junit-jupiter/5.21.0

AbgarSim replied:
Yeah, you are rightly deduced - need this dependency for `MockitoExtension.class`.
Double checked and sadly also couldn't find a managed version anywhere

AbgarSim commented on pom.xml:
```diff
@@ -111,9 +111,9 @@
-        <version>4.13.2</version>
+        <groupId>org.junit.jupiter</groupId>
+        <artifactId>junit-jupiter</artifactId>
+        <version>5.11.4</version>
```
So I'm not quite sure how this repo is linked to `gapic-generator-java-pom-parent` and just using `${junit.version}` fails locally. 

However, following the trail of parents of this repo i was able to reach native-image-shared-config pom which has `<junit-vintage-engine.version>5.11.4</junit-vintage-engine.version>`

I used this placeholder instead, please advice if this is acceptable or how to reach `${junit.version}`

AbgarSim commented:
Hi @lqiu96 thanks for the feedback!

I've pushed the following adjustments:

- Deleted `junit-platform.properties` - will move it as separate PR
- Adjusted all the `Assertion.*` imports, good catch missed those
- Used the `junit-vintage-engine.version` for junit dep - as per my comment above please advice if this is ok
- Split JUnit Jupiter into explicit `junit-jupiter-api` and `junit-jupiter-engine`, this should address the dependencies check

AbgarSim commented:
> Thanks, I took a look at the CI and I think there are a few minor things
> 
> * Linter is complaining, I think we just need to run `mvn fmt` again
> * Dependencies CI is complaining:
> 
> ```
> Error:  Unused declared dependencies found:
> Error:     org.junit.jupiter:junit-jupiter-engine:jar:5.11.4:test
> ```
> 
> Is it possible to remove `junit-jupiter-engine` from the pom? I think that should satisfy the CI (it may also be a false flag)
> 
> * CI is complaining about Mockito 5.x requiring Java 11 minimum for the Java 8 test. Apologies, can we downgrade this to mockito v4.11.0? It's a test dependency, but we haven't split the source JDK requirement version and from any test JDK requirements.
> 
> Changes LGTM. I'll approve after these which I think should fix everything. Thanks again!

This is done
- Ran the ` mvn fmt:format -T 1C ` command 
- Removed the `junit-jupiter-engine` from POM
- Downgraded mockito BOM to **4.11.0**

lqiu96 commented:
/gcbrun

lqiu96 commented:
Looks like Graalvm is complaining:
```
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 0, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] 
[INFO] --- native:0.10.6:test (test-native) @ google-cloud-bigquery ---
[ERROR] Test configuration file wasn't found. Make sure that test execution wasn't skipped.
```

Not sure why. Taking a look...

lqiu96 commented:
Thanks @AbgarSim, this is done in https://github.com/googleapis/java-bigquery/pull/4052

AbgarSim commented:
> Thanks @AbgarSim, this is done in #4052

Thanks! Created follow up in this https://github.com/googleapis/java-bigquery/pull/4058
